services/wallet/router/router_v2_test.go

```diff
@@ -0,0 +1,243 @@
+package router
+
+import (
+	"errors"
+	"math/big"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	walletCommon "github.com/status-im/status-go/services/wallet/common"
+	"github.com/status-im/status-go/services/wallet/router/pathprocessor"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestValidateInputData(t *testing.T) {
+	testCases := []struct {
+		name          string
+		input         *RouteInputParams
+		expectedError error
+	}{
+		{
+			name: "ENSRegister valid data on testnet",
+			input: &RouteInputParams{
+				SendType:    ENSRegister,
+				Username:    "validusername.eth",
+				PublicKey:   "validpublickey",
+				TokenID:     pathprocessor.SttSymbol,
+				TestnetMode: true,
+			},
+			expectedError: nil,
+		},
+		{
+			name: "ENSRegister valid data on mainnet",
+			input: &RouteInputParams{
+				SendType:  ENSRegister,
+				Username:  "validusername.eth",
+				PublicKey: "validpublickey",
+				TokenID:   pathprocessor.SntSymbol,
+			},
+			expectedError: nil,
+		},
+		{
+			name: "ENSRegister missing username",
+			input: &RouteInputParams{
+				SendType:    ENSRegister,
+				PublicKey:   "validpublickey",
+				TokenID:     pathprocessor.SttSymbol,
+				TestnetMode: true,
+			},
+			expectedError: errors.New("username and public key are required for ENSRegister"),
```

Comment:
These tests are very clear!

----

Going on a tangent, just to touch again on this topic because we talked about it a long time ago. I was thinking now about how clients will handle router errors. Nowadays, we poorly handle errors in mobile, generally speaking, and it's very brittle if we need to do that anyway because the errors exposed by the backend don't have IDs and lack structure. Maybe for the router v2 we could think of ways to use a more future proof implementation and maybe set an example to be used in future implementations? I'm not thinking anything too complicated, just globally unique errors and with some consistent structure using the type system.

This kind of idea could also help us reduce the brittleness and duplication in tests that need to match directly against error strings.

Reply:
I agree, we can have a list of errors as constants, and return a unique constant for each error.
Or returning something like this:
```
type ErrorResponse struct {
    Error  string `json:"error"`
    Code   int    `json:"code"`
    Detail string `json:"detail,omitempty"`
}
```

Brief context about the router v2, it has the following logic parts in terms of errors:
- validating input parameters, means lacking any mandatory field or inappropriately set field will result in an error
- errors that occur in the "searching for candidate" part of the algorithm are usually ignored in a way that we simply skip a certain transaction type if there is an error, that's considered as not possible path for the input parameters
- errors that may occur in the last part of the algorithm, after we get the best route, refer to if the user has enough assets to proceed with the suggested route or not